File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/StringLiteralShouldNotBeDuplicated.cs
using System;

namespace Tests.Diagnostics
{
    public class Program
    {
        public const string NameConst = "foobar"; // Noncompliant {{Define a constant instead of using this literal 'foobar' 8 times.}}
//                                      ^^^^^^^^
        public static readonly string NameReadonly = "foobar";
//                                                   ^^^^^^^^ Secondary


        private string name = "foobar";
//                            ^^^^^^^^ Secondary

        private string[] values = new[] { "something", "something", "something" }; // Compliant - repetition below threshold

        private string Name { get; } = "foobar";
//                                     ^^^^^^^^ Secondary

        public Program()
        {
            var x = "foobar";
//                  ^^^^^^^^ Secondary

            var y = "FooBar"; // Compliant - casing is different
        }

        public void Do(string s = "foobar")
//                                ^^^^^^^^ Secondary
        {
            var x = s ?? "foobar";
//                       ^^^^^^^^ Secondary

            string GetFooBar()
            {
                return "foobar";
//                     ^^^^^^^^ Secondary
            }
        }

        public void Validate(object foobar)
        {
            if (foobar == null)
            {
                throw new ArgumentNullException("foobar"); // Compliant - matches one of the parameter name
            }

            Do("foobar"); // Compliant - matches one of the parameter name
        }
    }

    public struct Foo
    {
        public const string NameConst = "foobar"; // Noncompliant {{Define a constant instead of using this literal 'foobar' 4 times.}}
//                                      ^^^^^^^^

        private string name1 = "foobar"; // Secondary
        private string name2 = "foobar"; // Secondary
        private string name3 = "foobar"; // Secondary
    }

    public class OuterClass
    {
        private string Name { get; } = "foobar"; // Noncompliant

        private class InnerClass
        {
            private string name1 = "foobar"; // Secondary - inner class count with base
            private string name2 = "foobar"; // Secondary
            private string name3 = "foobar"; // Secondary
            private string name4 = "foobar"; // Secondary
        }

        private struct InnerStruct
        {
            private string name1 = "foobar"; // Secondary - inner struct count with base
            private string name2 = "foobar"; // Secondary
            private string name3 = "foobar"; // Secondary
            private string name4 = "foobar"; // Secondary
        }
    }
}

File: analyzers/tests/SonarAnalyzer.Rules.Tests/Cases/StringLiteralShouldNotBeDuplicated.vb
Namespace Tests.TestCases

    Public Class Program

        Private empty As String = ""

        Public const NameConst As String = "foobar" ' Noncompliant {{Define a constant instead of using this literal 'foobar' 7 times.}}
        '                                  ^^^^^^^^
        Public ReadOnly NameReadonly As String = "foobar"
        '                                        ^^^^^^^^ Secondary

        Private name As String = "foobar"
        '                        ^^^^^^^^ Secondary

        Private values As String() = { "something", "something", "something" } ' Compliant - repetition below threshold

        Private ReadOnly Property FirstName = "foobar"
        '                                     ^^^^^^^^ Secondary

        Public Sub New()

            Dim x As String = "foobar"
            '                 ^^^^^^^^ Secondary

            Dim y As String = "FooBar" ' Compliant - casing Is different

        End Sub

        public Sub DoSomething(Optional s As String = "foobar")
        '                                             ^^^^^^^^ Secondary
            Dim x As String = If(s, "foobar")
            '                       ^^^^^^^^ Secondary
        End Sub

        public Sub Validate(fOobAR As Object)
             If foobar Is Nothing Then Throw New System.ArgumentNullException("foobar") ' Compliant - matches one of the parameter name

            DoSomething("foobar") ' Compliant - matches one of the parameter name

        End Sub

    End Class

    Public Class OuterClass

        Private ReadOnly Property Name As String = "foobar" ' Noncompliant

        Private Class InnerClass

            Private name1 As String = "foobar" ' Secondary - inner class count with base
            Private name2 As String = "foobar" ' Secondary
            Private name3 As String = "foobar" ' Secondary
            Private name4 As String = "foobar" ' Secondary

        End Class

        Private Structure InnerStruct

            private name1 As String
            private name2 As String
            private name3 As String
            private name4 As String

            public Sub New(s As String)

                name1 = "foobar" ' Secondary - inner struct count with base
                name2 = "foobar" ' Secondary
                name3 = "foobar" ' Secondary
                name4 = "foobar" ' Secondary

            End Sub

        End Structure

    End Class

    Public Structure OuterStruct

        Private Name As String
        Public Sub New(s As String)

            Name = "foobar" ' Noncompliant

        End Sub

        Private Structure InnerStruct
        
            private name1 As String
            private name2 As String
            private name3 As String
            private name4 As String

            public Sub New(s As String)
                name1 = "foobar" ' Secondary - inner struct count with base
                name2 = "foobar" ' Secondary
                name3 = "foobar" ' Secondary
                name4 = "foobar" ' Secondary
            End Sub

        End Structure

    End Structure

End Namespace
